### ISSUE TITLE
Add namespacing

### ISSUE THREAD

**Issue by lefticus** [ChaiScript/ChaiScript#113]
ChaiScript needs namespace support for various reasons: general C++ compatibility, static functions, namely.

**Comment by jeaye** [comment]
I found myself looking for namespaces just now. I was hoping that Modules just had a name, so I could make a Module called "Math", for example, and put my shit in there.

**Comment by lefticus** [comment]
I'm starting to think this is actually the next highest priority need for ChaiScript.

**Comment by gfannes** [comment]
Namespace support would be great indeed! Maybe adding support for statics and allowing a class to be reopened would even be enough.

**Comment by therocode** [comment]
As I'm making a dynamic content system where there is no way to know what script files will be added in advance from various sources, not being able to have a form of namespacing gravely risks collisions between difference scripts. I too would vote for this one as high prio.

**Comment by roig** [comment]
I think this should be on high priority too. Right now I need them hehe.

**Comment by lefticus** [comment]
Let's discuss what this should look like. Right now you can use a dynamic object to emulate a namespace similar to what javascript would do.

For example:

```
global mynamespace = Dynamic_Object();
mynamespace.function1 = fun(){ /* do something */ };
```

then:

```
mynamespace.function1();
```

thoughts on this?

**Comment by PhilCK** [comment]
Hello I was just looking at namespaces to impliment callbacks from C++ land.

I don't necessarily have a preference, so I'll just present this use-case. 

```
def on_start() // Clients function.
{
}

def Foo.on_start() // C++ callback
{
}
```

Quite like the way Lua does this.

```lua
Foo = {}

function Foo.my_hook()
end
```

I have no issue with implementing the callback as `foo_on_start()`, but this was why I just googled 'Chaiscript namespaces'

**Comment by CTrauma** [comment]
I've gotten my project to the point where not having namespaces is blocking my progress so I have been looking at dynamic objects to emulate namespaces. I had previously been using pybind11 with modules to simulate namespaces with my python bindings. The only real example I have found of using dynamic objects is https://github.com/ChaiScript/ChaiScript/blob/develop/samples/example.cpp#L167 is there an example of adding to an already existing type. I'm thinking of the situation where say a C++ class defines a internal enum that needs to be exposed to chaiscript?

**Comment by lefticus** [comment]
That's a bad example for actually wanting to use Dynamic_Object from C++. That example.cpp is actually very old and has some bad practices in it, but it's kept around to make sure the C++ API doesn't stray too far from the original intent.

So, I believe what you would want to do would be something like this:

```
// c++
chaiscript::dispatchkit::Dynamic_Object o;
o["value1"] = chaiscript::const_var(1);
o["value2"] = chaiscript::const_var(2);
o["some_func"] = chaiscript::var(chaiscript::fun([](const int lhs, const int rhs){ return lhs + rhs }));
chai.add_global(chaiscript::var(o), "MyNamespace");
```

```
// chaiscript
MyNamespace.value1 // returns 1
MyNamespace.value2 // returns 2
MyNamespace.some_func(3, 4) // returns 7
MyNamespace.some_new_func = fun(lhs, rhs) { lhs * rhs; } // adds new func to namespace
MyNamespace.some_new_func(2, 10) // returns 20
```

Let me know if this makes sense and ends up working out for you. If it does, it's something I can start formalizing into something more generally usable.

**Comment by CTrauma** [comment]
sorry didn't get back sooner, I started to look at this but another fire pulled me away. Thanks for the info, I'll try it out and let you know how it works out.

**Comment by lefticus** [comment]
I've been reluctant for years to just slap in something we might regret in the future - so if you can give some practical experience, that would be great. Thanks for taking the time to respond.

**Comment by stephenberry** [comment]
I've taken a look at how namespaces could be handled in ChaiScript.
@lefticus I started with your Dynamic_Object example in the comments above and will present my thoughts on this approach.

#### In General
- I strongly believe that namespaces should be objects. I think this approach will be intuitive to the user, flexible, and straightforward to implement with the existing code base.

#### A Simple Importer
This importer makes it possible to import specific namespaces per instance of ChaiScript.  
If multiple instances of ChaiScript were to be handled in an application, this could be used to only load the specific functions needed per instance.

``` C++
using namespace chaiscript;
std::shared_ptr<ChaiScript> chai = makeChaiScript();

// Create a map of namespace names with a Dynamic_Object for each namespace
std::shared_ptr<std::map<std::string, dispatch::Dynamic_Object>> chai_namespaces = std::make_shared<std::map<std::string, dispatch::Dynamic_Object>>();

// Add a "math" namespace with a constant and a function
dispatch::Dynamic_Object& math = (*chai_namespaces)["math"]; // The access ["math"] will instantiate a Dynamic_Object within the chai_namespaces
constexpr double pi = 3.14159265358979;
math["pi"] = const_var(pi);
math["sin"] = var(fun([](const double x) { return sin(x); }));

// Add a function for importing namespaces
chai->add(fun([=](const std::string& namespace_name) { chai->add_global(var((*chai_namespaces)[namespace_name]), namespace_name); }), "import");
```

#### Using The Importer (ChaiScript)
```
import("math")

print(math.pi) // prints 3.14159
print(math.sin(math.pi / 2.0)) // prints 1
```
* I really like how clean this syntax is. So, the handling of namespaces from the C++ side is nearly already available. It may be useful to create some helper classes. **However, the syntax for handling namespaces from the C++ side should integrate well with creating and using namespaces within ChaiScript.**

* An improvement to this importer would be to add a generic way of importing just a single function or a variable from a namespace:
``` import("math.pi") // only imports the "pi" constant from the math namespace ```
This improvement should be straightforward to implement.

* Another significant improvement would be to build the ```chai_namespaces``` map at compile time. This would reduce application load time. But, I'm not sure if this could be done in a straightforward and clean way. I'm really looking forward to constexpr lambda functions in C++17.

#### Namespace Generation in ChaiScript
Let's add a function called "namespace" that can create new namespaces and will register an "add" function with that namespace for adding objects/functions to the new namespace.
* Ideally this ```namespace``` function would add namespaces to the ```chai_namespaces``` map. But, for simplicity this code is just registering the namespace with the current instance of ChaiScript.
``` C++
chai->add(fun([=](const std::string& namespace_name)
{
   std::shared_ptr<dispatch::Dynamic_Object> namespace_object = std::make_shared<dispatch::Dynamic_Object>();

   (*namespace_object)["add"] = var(fun([=](const std::string& func_name, const Boxed_Value& boxed_value) { (*namespace_object)[func_name] = boxed_value; }));

   chai->add_global(var(namespace_object), namespace_name); return namespace_object; }) // We are returning the namespace_object so that we can immediately append to it. For example: namespace("new_namespace_name").add("x", 5.0)
      , "namespace");
```

#### Using The "namespace" Function (ChaiScript)
```
import("math")

namespace("trig") // define a new namespace called "trig"
trig.add("pi", math.pi) // add a variable to the trig namespace, in this case we're using pi from the math namespace
print(trig.pi)

trig.add("x", 5.0)
print(trig.x)

def myfun(x) { math.sin(x) } // create a new function that uses the math.sin function
trig.add("sin", myfun) // add myfun to the trig namespace, but call it "sin"
print(trig.sin(math.pi / 2.0))
```
* This all works just fine. But, I don't like the syntax for adding new variables and functions to the ChaiScript generated namespace.

#### Issues With ChaiScript Generated Namespace Syntax
* I don't like having to declare my function before I define its namespace.
* I don't like having to define the name of the function or variable when it already has a name. _Though this could be useful for calling the same function something else in a different namespace (e.g. changing "myfun" to "sin" for the trig namespace)._
* I don't like using parenthesis as it looks too much like an ordinary function call.

#### Ideal Syntax and Recommendations
Defining namespace for a variable: ```var math.tau = 5.0```  
Defining namespace for a function: ```def math.squared(x) { x*x }```

* These ideal examples would require changes to ChaiScript, but I feel they conform best with the semantics of ChaiScript.

* ```math``` as a namespace would probably need to be a special namespace structure so that the scripting engine could know when the access operator (.) is being called on a namespace and not trying to access a missing parameter. For namespaces the access operator (.) would create a new object if an object of that name doesn't exist. If an object of the same name already exists, then ChaiScript would throw an already defined error.

* I'm not sure how I feel about a "using namespace" mechanism. I think it could be handled when ```import``` is called, but I am wary about allowing namespaces to be removed in the middle of code or within specific scopes.

My last thought is that ChaiScript should reserve the word "namespace" and perhaps the word "import", as I would recommend these terms for adding namespaces and importing namespaces/modules.

**Comment by stephenberry** [comment]
Wow! I didn't realize how powerful the Dynamic_Object is until I was looking at more documentation.

A good chunk of what I wrote above is irrelevant and already supported.

The following code works just great in ChaiScript:
```
namespace("trig")

trig.pi = math.pi // add a variable to the trig namespace
print(trig.pi)

trig.sin = fun(x) { math.sin(x) }
print(trig.sin(math.pi / 2.0)) // prints 1
```
So, this form of namespace handling is already available.

@lefticus If you think handling namespaces like this is a good idea, then I can write some code for ```namespace``` and ```import``` methods and helper functions to be added to ChaiScript.

**Comment by lefticus** [comment]
Sure, give it a shot and let me know what you come up with, submit a pull request.

I'll eventually want to add some syntactic sugar around it so you could do something like:

```
namespace trig {
  var pi = 3.1;
}
```

or something.

**Comment by stephenberry** [comment]
I submitted a pull request to the develop branch with fundamental namespace support.

### Basic Syntax (C++)
``` C++
using namespace chaiscript;
Namespace math;
constexpr double pi = 3.14159265358979;
math["pi"] = const_var(pi);
math["tau"] = const_var(2.0*pi);
math["sin"] = var(fun([](const double x) { return sin(x); }));
chai->register_namespace(math, "math");
```
* `chaiscript::Namespace` is an alias for `dispatch::Dynamic_Object`, so that the user can write cleaner and more explicit code.
* `register_namespace` does not add the namespace to the ChaiScript instance, but simply registers it, so that it can be imported.

To import a namespace via C++ (_not generally recommended_):
``` C++
chai->import("math");
```
* If a namespace has been imported, subsequent calls to import the same namespace are ignored. Similar to including headers only once.

### Delayed Namespace Generation (C++)
If namespaces are large, significant memory and computing costs can be saved by not instantiating the namespace until it is imported.
``` C++
chai->register_namespace([]() {
      Namespace math;
      constexpr double pi = 3.14159265358979;
      math["pi"] = const_var(pi);
      math["tau"] = const_var(2.0*pi);
      math["sin"] = var(fun([](const double x) { return sin(x); }));
      return math; },
      "math");
```
* Passing a lambda function that returns a namespace will delay the namespace generation until `import` is called.

Example of large memory savings if the namespace `numbers` is not imported:
``` C++
chai->register_namespace([]() {
      Namespace numbers;
      for (size_t i = 0; i < 100000; ++i)
         numbers["num" + std::to_string(i)] = const_var(i);
      return numbers; },
      "numbers");
```

### ChaiScript Syntax
```
import("math")

print(math.pi)
print(math.sin(math.pi / 2.0))

namespace("trig")

trig.pi = math.pi

print(2.0*trig.pi)

trig.x = 5.0
print(trig.x)

trig.sin = fun(x) { math.sin(x) }
print(trig.sin(math.pi / 2.0))
```

### Global Namespaces
Namespaces are imported globally so that they can be used within the scope of function/class definitions:
```
def sin(x) { math.sin(x) }
```

### Concluding Remarks
* I attempted to implement a means of importing only a single element from a namespace: `import("math.pi")`. However, this doesn't make sense in the context of using Dynamic_Object for namespaces. I would have to make a new Dynamic_Object named `math` that only had an instance of `pi`. Because the Dynamic_Object can change, this new `math` object could become something totally different from the originally registered `math` namespace. Therefore, the current implementation doesn't allow partial namespace inclusion, and I'm not sure it should, unless a significantly different form of namespace handling is developed.

* I haven't looked at implementing nested namespaces. This could be a future improvement.

* I think the ability to delay generating a namespace (essentially a library), will be a huge benefit to ChaiScript, as it allows a program to have a huge library of functions/objects available for scripting, yet only the used parts of the library are loaded.

**Comment by totalgee** [comment]
Is there a way you can define a `user_type` in a "namespace" this way (i.e. using a `Dynamic_Object` as the namespace) without also defining it globally, at the "top level", as a type?

It seems to _almost_ work.  The problem I see is defining multiple (overloaded) methods with the same name, since the Dynamic_Object doesn't have an add() method that supports multiple versions (as `Module` does) with the same name.  But it's quite close; the following works:

```c++
    auto ns = std::make_shared<dispatch::Dynamic_Object>();
    gChai->add_global(var(ns), "math");
    (*ns)["vec3"] = var(user_type<glm::vec3>());
    (*ns)["vec3"] = var(constructor<glm::vec3(glm::vec3 const&)>()); // copy constructor
    (*ns)["vec3"] = var(constructor<glm::vec3(float,float,float)>());
    (*ns)["vec3"] = var(constructor<glm::vec3()>()); // default constructor
    (*ns)["="] = var(fun(static_cast<glm::vec3& (glm::vec3::*)(glm::vec3 const&)>(&glm::vec3::operator=))); // assignment operator
```

After doing this, we can successfully create an instance of the "namespaced" object in scripting, using `var v = math.vec3()`, and we can even assign one math.vec3 to another using `operator=`, but we can't access the copy constructor or 3-float constructor, which have been overwritten with the default constructor (i.e. the last one set using the same name in the `m_attrs` map wins).

Any way this could be achieved?  Or is the only way to define the user_types globally, then make an "alias" pointing to it from a different namespace?

**Comment by thomasfannes** [comment]
I would like to reopen the question from @totalgee: Isn't there away that namespaces can be more like Modules? This would make it possible that a namespace has
 * function overload
 * User types
 * Can be created by combining different modules

in stead of Namespace being a dynamic object, could it not be a module?

**Comment by lefticus** [comment]
It would be best if you open a new issue that shows what needs to be different from the namespace support that was already merged last year. @thomasfannes and discusses what the issue is

**Comment by daneldanel** [comment]
Please provide a sample how to define a class in a namespace from c++ usable from chaiscript
`namespace ns {
class myclass {
public:
    myclass() {}
};
};`

in chai script:
import("ns");
auto x = ns.myclass();

**Comment by stellarpower** [comment]
As a newcomer to the toolkit, I'd echo what has been said above. Using the DynamicObject seems a really flexible way of doing thingsm I like it to an extent, but it has been a bit confusing for me that there are modules and namespaces, and that the two seem quite different. I want to expose some types from OpenCV. I'd like to be able to add those simply using `add_class`, but it would also be helpful to keep them in a separate scope, away from my user types in the global namespace. It's not a must, and admittedly I don't know anything baout the implementation, but from a user-programmer perspective, I think it would simplify the design if it were possible to unite modules and namespaces into the same concept. For example, if something like `chai.add(myModule, "name")` were possible, it could introduce the module as a separate namespace, whilst `chai.add(myModule)` would continue the existing module behaviour, importing the members of the module into the global namespace. The lambda callback continues to make sense, and I agree is an easy and minimal way of speeding up the runtime performance, if I don't call `import("name");` from the script. I don't know if deferred evaluation is possible at all in a general sense, for situations like `chai.add()`. I.e. if we can avoid bothering to populate a symbol table fully with overloads for some function, if this function is never actually called from the script - instead just inseritng a reference that they may be evaluated at a later time.
